lib/storage.hh

```diff
@@ -0,0 +1,456 @@
+      notice, this list of conditions and the following disclaimer.
+
+    * Redistributions in binary form must reproduce the above
+      copyright notice, this list of conditions and the following
+      disclaimer in the documentation and/or other materials provided
+      with the distribution.
+
+    * Neither the name of the University of California nor the names
+      of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written
+      permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+LICENSE (END)
+
+Contact: khmer-project@idyll.org
+*/
+
+#ifndef STORAGE_HH
+#define STORAGE_HH
+
+namespace khmer
+{
+typedef std::map<HashIntoType, BoundedCounterType> KmerCountMap;
+class Hashbits;
+class CountingHash;
+
+//
+// base Storage class for hashtable-related storage of information in memory.
+//
+
+class Storage {
+public:
+    bool _use_bigcount;
+
+    Storage() : _use_bigcount(false) { } ;
+    virtual ~Storage() { }
+    virtual std::vector<uint64_t> get_tablesizes() const = 0;
+    virtual const size_t n_tables() const = 0;
+    virtual void save(std::string, WordLength) = 0;
+    virtual void load(std::string, WordLength&) = 0;
+    virtual const uint64_t n_occupied() const = 0;
+    virtual const uint64_t n_unique_kmers() const = 0;
+    virtual BoundedCounterType test_and_set_bits( HashIntoType khash ) = 0;
+    virtual void add(HashIntoType khash) = 0;
+    virtual const BoundedCounterType get_count(HashIntoType khash) const = 0;
+    virtual Byte ** get_raw_tables() = 0;
+
+    void set_use_bigcount(bool b);
+    bool get_use_bigcount();
+};
+
+
+/*
+ * \class BitStorage
+ *
+ * \brief A Bloom filter implementation.
+ *
+ * BitStorage is used to track presence/absence of k-mers by Hashtable
+ * and derived classes.  It contains 'n_tables' different tables of
+ * bitsizes specified in 'tablesizes' (so 1/8 for bytesizes).
+ *
+ * Like other Storage classes, BitStorage manages setting the bits and
+ * tracking statistics, as well as save/load, and not much else.
+ *
+ */
+
+class BitStorage : public Storage
+{
+friend class Hashbits;
+protected:
+    std::vector<uint64_t> _tablesizes;
+    size_t _n_tables;
```

Comment:
Isn't this redundant with `_tablesizes.size()`? Not a big deal, but why risk the two variables getting out of sync?